File: src/main/java/dev/slne/discord/listener/whitelist/WhitelistJoinListener.java
package dev.slne.discord.listener.whitelist;

import dev.slne.discord.annotation.DiscordListener;
import dev.slne.discord.config.discord.GuildConfig;
import dev.slne.discord.spring.feign.dto.WhitelistDTO;
import dev.slne.discord.spring.service.whitelist.WhitelistService;
import javax.annotation.Nonnull;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;

/**
 * The type WhitelistDTO join listener.
 */
@DiscordListener
public class WhitelistJoinListener extends ListenerAdapter {

  private final WhitelistService whitelistService;

  @Autowired
  public WhitelistJoinListener(WhitelistService whitelistService) {
    this.whitelistService = whitelistService;
  }

  @Override
  public void onGuildMemberJoin(@Nonnull GuildMemberJoinEvent event) {
    handleEvent(event);
  }

  @Async
  protected void handleEvent(GuildMemberJoinEvent event) {
    final User user = event.getUser();
    final Guild guild = event.getGuild();
    final WhitelistDTO whitelist = whitelistService.getWhitelistByDiscordId(user.getId()).join();

    if (whitelist == null) {
      return;
    }

    final GuildConfig guildConfig = GuildConfig.getConfig(guild.getId());
    if (guildConfig == null) {
      return;
    }

    final Role whitelistedRole = guildConfig.getWhitelistedRole();
    if (whitelistedRole == null) {
      return;
    }

    final Member member = guild.retrieveMember(user).complete();
    if (member == null) {
      return;
    }

    guild.addRoleToMember(member, whitelistedRole).queue();
  }
}

File: src/main/kotlin/dev/slne/discord/listener/whitelist/WhitelistJoinListener.kt
package dev.slne.discord.listener.whitelist

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.events.listener
import dev.slne.discord.guild.getDiscordGuildByGuildId
import dev.slne.discord.persistence.service.whitelist.WhitelistService
import jakarta.annotation.PostConstruct
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent
import net.kyori.adventure.text.logger.slf4j.ComponentLogger
import org.springframework.stereotype.Component

@Component
class WhitelistJoinListener(private val jda: JDA, private val whitelistService: WhitelistService) {

    private val logger = ComponentLogger.logger()

    @PostConstruct
    fun registerListener() {
        jda.listener<GuildMemberJoinEvent> { event ->
            val user = event.user
            val guild = event.guild
            val whitelist = whitelistService.findByDiscordId(user.id) ?: return@listener

            whitelist.blocked = false
            whitelistService.saveWhitelist(whitelist)

            val guildConfig = getDiscordGuildByGuildId(guild.id)?.discordGuild ?: return@listener
            val whitelistedRole = guildConfig.whitelistRole ?: return@listener
            val member = guild.retrieveMember(user).await() ?: return@listener

            guild.addRoleToMember(member, whitelistedRole).await()

            logger.info("User ${user.name} left the server and was blocked.")

            user.openPrivateChannel().await()
                .sendMessage("Da du dich auf der Whitelist für den CastCrafter Community Server befindest, wurde deine Whitelist entblockt und dir die Rolle \"${whitelistedRole.name}\" wieder zugeteilt! Viel Spaß auf dem Server!")
                .await()
        }
    }
}
